go-controller/pkg/ovn/zone_interconnect/chassis_handler.go

```diff
@@ -0,0 +1,188 @@
+	chassisID, err := util.ParseNodeChassisIDAnnotation(node)
+	if err != nil {
+		return fmt.Errorf("failed to parse node chassis-id for node - %s, error: %w", node.Name, err)
+	}
+
+	ch := &sbdb.Chassis{
+		Name:     chassisID,
+		Hostname: node.Name,
+	}
+
+	chassis, err := libovsdbops.GetChassis(zch.sbClient, ch)
+	if err != nil {
+		return fmt.Errorf("failed to get the chassis record for the local zone node %s, error: %w", node.Name, err)
+	}
+
+	// Update the node chassis to local if its not explicitly marked as local
+	if chassis.OtherConfig != nil {
+		if strings.ToLower(chassis.OtherConfig["is-remote"]) == "false" {
+			// Nothing to do. The chassis is already marked as local.
+			return nil
+		}
+		chassis.OtherConfig["is-remote"] = "false"
+	} else {
+		chassis.OtherConfig = map[string]string{
+			"is-remote": "false",
+		}
+	}
+
+	if err := libovsdbops.UpdateChassisOtherConfig(zch.sbClient, chassis); err != nil {
+		return fmt.Errorf("failed to update chassis to local for node %s, error: %w", node.Name, err)
+	}
+
+	return nil
+}
+
+// AddRemoteZoneNode creates the remote chassis for the remote zone node node in the SB DB or marks
+// the entry as remote if it was local chassis earlier.
+func (zch *ZoneChassisHandler) AddRemoteZoneNode(node *corev1.Node) error {
+	// Get the chassis id.
+	chassisID, err := util.ParseNodeChassisIDAnnotation(node)
+	if err != nil {
+		return fmt.Errorf("failed to parse node chassis-id for node - %s, error: %w", node.Name, err)
+	}
+
+	nodePrimaryIp, err := util.GetNodePrimaryIP(node)
+	if err != nil {
+		return fmt.Errorf("failed to parse node %s primary IP %w", node.Name, err)
+	}
+
+	ch := &sbdb.Chassis{
+		Name:     chassisID,
+		Hostname: node.Name,
+	}
+	chassis, err := libovsdbops.GetChassis(zch.sbClient, ch)
+	if err != nil {
+		if !errors.Is(err, client.ErrNotFound) {
+			return err
+		}
+
+		// Remote chassis record not found. Create it.
+		chassis = &sbdb.Chassis{
+			Name:     chassisID,
+			Hostname: node.Name,
+			OtherConfig: map[string]string{
+				"is-remote": "true",
+			},
+		}
+		encap := &sbdb.Encap{
+			ChassisName: chassisID,
+			IP:          nodePrimaryIp,
+			Type:        "geneve",
+			Options:     map[string]string{"csum": "true"},
+		}
+		if err := libovsdbops.CreateChassis(zch.sbClient, chassis, encap); err != nil {
+			return fmt.Errorf("failed to create encaps and remote chassis in SBDB for node %s, error: %w", node.Name, err)
+		}
+	} else {
+		if chassis.OtherConfig != nil {
+			if strings.ToLower(chassis.OtherConfig["is-remote"]) == "true" {
+				// Nothing to do. The chassis is already marked as remote.
+				return nil
+			}
+			chassis.OtherConfig["is-remote"] = "true"
+		} else {
+			chassis.OtherConfig = map[string]string{
+				"is-remote": "true",
+			}
+		}
+		if err := libovsdbops.UpdateChassisOtherConfig(zch.sbClient, chassis); err != nil {
+			return fmt.Errorf("failed to update the remote chassis in SBDB for node %s, error: %w", node.Name, err)
+		}
+	}
```

Comment:
nit: honestly, this would all just be much simpler if we just had a `CreateOrUpdateChassis` in libovsdbops

For the the update path specifically, the chassis is read from cache two times.

Reply:
When a chassis has to be created we need to create Encaps too.  But if a chassis entry is already present, we just need to update the chassis .

How to do that in CreateOrUpdateChassis() ?  I could not figure out.

Reply:
If suppose we have a function CreateOrUpdateChassis(), we need to query first for the existence of the chassis record.  Without that we will end up creating another encaps record resulting in constraint violation.

That's why I added one function - CreateChassis() which takes encaps info too.
the other one - UpdateChassisOtherConfig which just updates the otherconfig map.

Reply:
This would be the code:
```
func CreateOrUpdateChassis(sbClient libovsdbclient.Client, chassis *sbdb.Chassis, encap *sbdb.Encap) error {
	m := newModelClient(sbClient)
	opModels := []operationModel{
		{
			Model: encap,
			DoAfter: func() {
				chassis.Encaps = []string{encap.UUID}
			},
			ErrNotFound:    false,
			BulkOp:         false,
		},
		{
			Model:          chassis,
			ErrNotFound:    false,
			BulkOp:         false,
		},
	}

	if _, err := m.CreateOrUpdate(opModels...); err != nil {
		return err
	}
	return nil
}
```

Now what this piece of code does is:

1. If the same encap exists (one with either the same UUID, or if no UUID was provided, with the same type/IP pair), the provided information is updated for that encap. If the same encap does not exist, in gets created with the provided information.
2. If the same chassis exists (one with either the same UUID, or if no UUID was provided, with the same name), the provided information is updated for that chassis. If the same chassis does not exist, in gets created with the provided information. Please note that if chassis is associated with a different encap than it was before, the older encap will be garbage collected if it is not used anywhere else.

Note that as long as you are not trying to remove keys from encaps Options or chassis Otherconfig (you aren't), you don't need to read/get chassis at all either before calling CreateOrUpdate/CreateOrUpdateChassis. ModelClient is able to update keys values because it issues a delete for them before the insert (and it should handle just removing keys as well but we forgot about that, it should be easy to add).

Note also, that we are generally not concerned to avoid performing an update that does not really change anything and is just a noop. If we did, that could also be implemented generically within model client instead of having the caller be concerned with it but we haven't done that so far (assuming the DB server can do this more efficiently than us generally speaking). In this specific case is also of no concern as it would be unusual to be updating the chassis details.

Reply:
For the chassis encaps, you could do
```
chassis.Encaps = set.New(encap.UUID, chassis.Encaps...).UnsortedList()
```
That should ensure that you don't add the same UUID twice and make the method serve a more general purpose.

Reply:
You could use that `CreateOrUpdateChassis` instead of `UpdateChassisOtherConfig` as well. The only purpose of `UpdateChassisOtherConfig` now would be the ability to delete keys from OtherConfig so no problem having it. But you are not doing that and the ability to delete keys, as mentioned above, should rather be ingrained in mode client.

Reply:
I tried with your suggestion and it didn't work for me.

This is how it looks like in libovsdbops/chassis.go

```
// CreateOrUpdateChassis creates or updates the chassis record along with the encap record
func CreateOrUpdateChassis(sbClient libovsdbclient.Client, chassis *sbdb.Chassis, encap *sbdb.Encap, chassisFields ...interface{}) error {
	m := newModelClient(sbClient)
	if len(chassisFields) == 0 {
		chassisFields = onModelUpdatesAllNonDefault()
	}

	opModels := []operationModel{
		{
			Model: encap,
			DoAfter: func() {
				chassis.Encaps = []string{encap.UUID}
			},
			OnModelUpdates: onModelUpdatesNone(),
			ErrNotFound:    false,
			BulkOp:         false,
		},
		{
			Model:          chassis,
			OnModelUpdates: chassisFields,
			ErrNotFound:    false,
			BulkOp:         false,
		},
	}

	if _, err := m.CreateOrUpdate(opModels...); err != nil {
		return err
	}

	return nil
}

```

And in zone_interconnect/chassis_handler.go

```

// AddLocalZoneNode marks the chassis entry for the node in the SB DB to a local chassis
func (zch *ZoneChassisHandler) AddLocalZoneNode(node *corev1.Node) error {
	if err := zch.createOrUpdateNodeChassis(node, false); err != nil {
		return fmt.Errorf("failed to update chassis to local for local node %s, error: %w", node.Name, err)
	}

	return nil
}

// AddRemoteZoneNode creates the remote chassis for the remote zone node node in the SB DB or marks
// the entry as remote if it was local chassis earlier.
func (zch *ZoneChassisHandler) AddRemoteZoneNode(node *corev1.Node) error {
	if err := zch.createOrUpdateNodeChassis(node, true); err != nil {
		return fmt.Errorf("failed to create or update chassis to remote for remote node %s, error: %w", node.Name, err)
	}

	return nil
}


// createOrUpdateNodeChassis creates or updates the node chassis to local or remote.
func (zch *ZoneChassisHandler) createOrUpdateNodeChassis(node *corev1.Node, isRemote bool) error {
	// Get the chassis id.
	chassisID, err := util.ParseNodeChassisIDAnnotation(node)
	if err != nil {
		return fmt.Errorf("failed to parse node chassis-id for node - %s, error: %w", node.Name, err)
	}

	nodePrimaryIp, err := util.GetNodePrimaryIP(node)
	if err != nil {
		return fmt.Errorf("failed to parse node %s primary IP %w", node.Name, err)
	}

	chassis := sbdb.Chassis{
		Name:     chassisID,
		Hostname: node.Name,
		OtherConfig: map[string]string{
			"is-remote": strconv.FormatBool(isRemote),
		},
	}

	encap := sbdb.Encap{
		ChassisName: chassisID,
		IP:          nodePrimaryIp,
		Type:        "geneve",
		Options:     map[string]string{"csum": "true"},
	}

	return libovsdbops.CreateOrUpdateChassis(zch.sbClient, &chassis, &encap, &chassis.OtherConfig)
}
```

And I get the below error for the tests

```
I0419 10:22:10.589217 2151113 chassis.go:142] NUMAN : CreateOrUpdateChassis entered
I0419 10:22:10.589297 2151113 client.go:783]  "msg"="transacting operations" "database"="OVN_Southbound" "operations"="[{Op:insert Table:Encap Row:map[chassis_name:cb9ec8fa-b409-4ef3-9f42-d9283c47aac6 ip:10.0.0.10 options:{GoMap:map[csum:true]} type:geneve] Rows:[] Columns:[] Mutations:[] Timeout:<nil> Where:[] Until: Durable:<nil> Comment:<nil> Lock:<nil> UUIDName:u0685115531} {Op:update Table:Chassis Row:map[other_config:{GoMap:map[is-remote:true]}] Rows:[] Columns:[] Mutations:[] Timeout:<nil> Where:[where column _uuid == {cb9ec8fa-b409-4ef3-9f42-d9283c47aac6}] Until: Durable:<nil> Comment:<nil> Lock:<nil> UUIDName:}]"
2023/04/19 10:22:10 server/transaction/cache: "caller"={"file":"cache.go","line":1028} "level"=5 "msg"="processing update" "uuid"="fa7b9d2b-9d2f-40d9-97fa-c33ea96561b8" "table"="Encap"
2023/04/19 10:22:10 server/transaction/cache: "caller"={"file":"cache.go","line":1048} "level"=5 "msg"="inserting row" "uuid"="fa7b9d2b-9d2f-40d9-97fa-c33ea96561b8" "table"="Encap" "model"="&{UUID:fa7b9d2b-9d2f-40d9-97fa-c33ea96561b8 ChassisName:cb9ec8fa-b409-4ef3-9f42-d9283c47aac6 IP:10.0.0.10 Options:map[csum:true] Type:geneve}"
2023/04/19 10:22:10 server/transaction/cache: "caller"={"file":"cache.go","line":1028} "level"=5 "msg"="processing update" "uuid"="cb9ec8fa-b409-4ef3-9f42-d9283c47aac6" "table"="Chassis"
2023/04/19 10:22:10 server/transaction/cache: "caller"={"file":"cache.go","line":1069} "level"=5 "msg"="updated row" "uuid"="cb9ec8fa-b409-4ef3-9f42-d9283c47aac6" "table"="Chassis" "old"="&{UUID:cb9ec8fa-b409-4ef3-9f42-d9283c47aac6 Encaps:[] ExternalIDs:map[] Hostname:node1 Name:cb9ec8fa-b409-4ef3-9f42-d9283c47aac6 NbCfg:0 OtherConfig:map[is-remote:false] TransportZones:[] VtepLogicalSwitches:[]}" "new"="&{UUID:cb9ec8fa-b409-4ef3-9f42-d9283c47aac6 Encaps:[] ExternalIDs:map[] Hostname:node1 Name:cb9ec8fa-b409-4ef3-9f42-d9283c47aac6 NbCfg:0 OtherConfig:map[is-remote:true] TransportZones:[] VtepLogicalSwitches:[]}"
2023/04/19 10:22:10 server: "caller"={"file":"server.go","line":215} "msg"="Skipping transaction DB commit due to error" "error"="failed to process operation" "operations"=[{"op":"insert","table":"Encap","row":{"chassis_name":"cb9ec8fa-b409-4ef3-9f42-d9283c47aac6","ip":"10.0.0.10","options":{"GoMap":{"\"csum\"":"true"}},"type":"geneve"},"uuid-name":"fa7b9d2b-9d2f-40d9-97fa-c33ea96561b8"},{"op":"update","table":"Chassis","row":{"other_config":{"GoMap":{"\"is-remote\"":"true"}}},"where":["where column _uuid == {cb9ec8fa-b409-4ef3-9f42-d9283c47aac6}"]}] "results"=[{,"uuid":{"uuid":"fa7b9d2b-9d2f-40d9-97fa-c33ea96561b8"}},{"count":1,"uuid":{"uuid":""}},{,"error":"constraint violation","details":"operation would cause rows in the \"Encap\" table to have identical values (090c000667656e6576650c0c000931302e302e302e3130e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855) for index on column \"ip,type\". First row, with UUID fa7b9d2b-9d2f-40d9-97fa-c33ea96561b8, was inserted by this transaction. Second row, with UUID [4585e5bd-5c30-4181-b05a-70e8c6d42e3c], existed in the database before this operation and was not modified","uuid":{"uuid":""}}] "operation error"="constraint violation"
[91m[1m• Failure [0.234 seconds][0m
Zone Interconnect Chassis Operations
[90m/home/nusiddiq/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/zone_interconnect/chassis_handler_test.go:19[0m
  [91m[1mDelete remote zone node [It][0m
  [90m/home/nusiddiq/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/zone_interconnect/chassis_handler_test.go:191[0m

  [91mUnexpected error:
      <*fmt.wrapError | 0xc00043cee0>: {
          msg: "failed to create or update chassis to remote for remote node node1, error: error in transact with ops [{Op:insert Table:Encap Row:map[chassis_name:cb9ec8fa-b409-4ef3-9f42-d9283c47aac6 ip:10.0.0.10 options:{GoMap:map[csum:true]} type:geneve] Rows:[] Columns:[] Mutations:[] Timeout:<nil> Where:[] Until: Durable:<nil> Comment:<nil> Lock:<nil> UUIDName:u0685115531} {Op:update Table:Chassis Row:map[other_config:{GoMap:map[is-remote:true]}] Rows:[] Columns:[] Mutations:[] Timeout:<nil> Where:[where column _uuid == {cb9ec8fa-b409-4ef3-9f42-d9283c47aac6}] Until: Durable:<nil> Comment:<nil> Lock:<nil> UUIDName:}] results [{Count:0 Error: Details: UUID:{GoUUID:fa7b9d2b-9d2f-40d9-97fa-c33ea96561b8} Rows:[]} {Count:1 Error: Details: UUID:{GoUUID:} Rows:[]} {Count:0 Error:constraint violation Details:operation would cause rows in the \"Encap\" table to have identical values (090c000667656e6576650c0c000931302e302e302e3130e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855) for index on column \"ip,type\". First row, with UUID fa7b9d2b-9d2f-40d9-97fa-c33ea96561b8, was inserted by this transaction. Second row, with UUID [4585e5bd-5c30-4181-b05a-70e8c6d42e3c], existed in the database before this operation and was not modified UUID:{GoUUID:} Rows:[]}] and errors []: constraint violation: operation would cause rows in the \"Encap\" table to have identical values (090c000667656e6576650c0c000931302e302e302e3130e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855) for index on column \"ip,type\". First row, with UUID fa7b9d2b-9d2f-40d9-97fa-c33ea96561b8, was inserted by this transaction. Second row, with UUID [4585e5bd-5c30-4181-b05a-70e8c6d42e3c], existed in the database before this operation and was not modified",
          err: <*errors.errorString | 0xc00025ee00>{
              s: "error in transact with ops [{Op:insert Table:Encap Row:map[chassis_name:cb9ec8fa-b409-4ef3-9f42-d9283c47aac6 ip:10.0.0.10 options:{GoMap:map[csum:true]} type:geneve] Rows:[] Columns:[] Mutations:[] Timeout:<nil> Where:[] Until: Durable:<nil> Comment:<nil> Lock:<nil> UUIDName:u0685115531} {Op:update Table:Chassis Row:map[other_config:{GoMap:map[is-remote:true]}] Rows:[] Columns:[] Mutations:[] Timeout:<nil> Where:[where column _uuid == {cb9ec8fa-b409-4ef3-9f42-d9283c47aac6}] Until: Durable:<nil> Comment:<nil> Lock:<nil> UUIDName:}] results [{Count:0 Error: Details: UUID:{GoUUID:fa7b9d2b-9d2f-40d9-97fa-c33ea96561b8} Rows:[]} {Count:1 Error: Details: UUID:{GoUUID:} Rows:[]} {Count:0 Error:constraint violation Details:operation would cause rows in the \"Encap\" table to have identical values (090c000667656e6576650c0c000931302e302e302e3130e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855) for index on column \"ip,type\". First row, with UUID fa7b9d2b-9d2f-40d9-97fa-c33ea96561b8, was inserted by this transaction. Second row, with UUID [4585e5bd-5c30-4181-b05a-70e8c6d42e3c], existed in the database before this operation and was not modified UUID:{GoUUID:} Rows:[]}] and errors []: constraint violation: operation would cause rows in the \"Encap\" table to have identical values (090c000667656e6576650c0c000931302e302e302e3130e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855) for index on column \"ip,type\". First row, with UUID fa7b9d2b-9d2f-40d9-97fa-c33ea96561b8, was inserted by this transaction. Second row, with UUID [4585e5bd-5c30-4181-b05a-70e8c6d42e3c], existed in the database before this operation and was not modified",
          },
      }
      failed to create or update chassis to remote for remote node node1, error: error in transact with ops [{Op:insert Table:Encap Row:map[chassis_name:cb9ec8fa-b409-4ef3-9f42-d9283c47aac6 ip:10.0.0.10 options:{GoMap:map[csum:true]} type:geneve] Rows:[] Columns:[] Mutations:[] Timeout:<nil> Where:[] Until: Durable:<nil> Comment:<nil> Lock:<nil> UUIDName:u0685115531} {Op:update Table:Chassis Row:map[other_config:{GoMap:map[is-remote:true]}] Rows:[] Columns:[] Mutations:[] Timeout:<nil> Where:[where column _uuid == {cb9ec8fa-b409-4ef3-9f42-d9283c47aac6}] Until: Durable:<nil> Comment:<nil> Lock:<nil> UUIDName:}] results [{Count:0 Error: Details: UUID:{GoUUID:fa7b9d2b-9d2f-40d9-97fa-c33ea96561b8} Rows:[]} {Count:1 Error: Details: UUID:{GoUUID:} Rows:[]} {Count:0 Error:constraint violation Details:operation would cause rows in the "Encap" table to have identical values (090c000667656e6576650c0c000931302e302e302e3130e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855) for index on column "ip,type". First row, with UUID fa7b9d2b-9d2f-40d9-97fa-c33ea96561b8, was inserted by this transaction. Second row, with UUID [4585e5bd-5c30-4181-b05a-70e8c6d42e3c], existed in the database before this operation and was not modified UUID:{GoUUID:} Rows:[]}] and errors []: constraint violation: operation would cause rows in the "Encap" table to have identical values (090c000667656e6576650c0c000931302e302e302e3130e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855) for index on column "ip,type". First row, with UUID fa7b9d2b-9d2f-40d9-97fa-c33ea96561b8, was inserted by this transaction. Second row, with UUID [4585e5bd-5c30-4181-b05a-70e8c6d42e3c], existed in the database before this operation and was not modified
  occurred[0m
```

Reply:
I don't think CreateOrUpdateChassis() would work if the chassis entry already has an Encap set.

Reply:
I think that's because we should monitor the encap table in the client as well. See https://github.com/ovn-org/ovn-kubernetes/blob/a053d08ed6c900a38d3f6e7277e4ad9440f346aa/go-controller/pkg/libovsdb/libovsdb.go#L108

Reply:
Thanks a lot again for pointing to the right direction